openfermioncirq/variational/letter_with_subscripts.py

```diff
@@ -10,18 +10,46 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
-from typing import Union
+from typing import Union, List
 
-import cirq
+import sympy
 
 
-class LetterWithSubscripts(cirq.Symbol):
+def _name(letter: str, *subscripts: Union[str, int]) -> str:
+    return letter + ''.join('_{}'.format(subscript)
+                            for subscript in subscripts)
+
+
+class LetterWithSubscripts(sympy.Symbol):
+
+    def __new__(cls, letter: str, *subscripts: Union[str, int]):
+        return super().__new__(cls, _name(letter, *subscripts))
 
     def __init__(self,
                  letter: str,
                  *subscripts: Union[str, int]) -> None:
         self.letter = letter
         self.subscripts = subscripts
-        name = letter + ''.join('_{}'.format(subscript)
-                                for subscript in subscripts)
-        super().__init__(name)
+        super().__init__()
+
+    def __eq__(self, other):
```

Comment:
This equates `LetterWithSubscripts('x', 1, 2)` and `LetterWithSubscripts('x_1_2')`. If that's the case, why not just have a function `letter_with_subscripts` that returns a `sympy.Symbol` object directly?

Reply:
If we keep `LetterWithSubscripts`, it should at least validate its inputs. As is, it accepts empty strings and strings containing underscores as valid subscripts.

Reply:
@kevinsung has code in the library that uses the `.subscripts` and `.letter` properties in order to unpack this sorta-kinda symbol into other symbols. It's not ideal, but I didn't have anything better to suggest. Fixing it is beyond the scope of this PR.

Reply:
Agreed that removing it is out of scope.

In the meantime, it seems better that equality checks equality of the components rather than just the overall string.

@kevinsung Where are these uses of the properties?

Reply:
Messing with the equality will create weird interactions with sympy, which probably assumes in various places that symbol equality acts like string equality. The fact that I had to sneakily override a method in order to make substitution work indicates they did not design the class with the intention of it being overriden.

Reply:
Okay, I don't want to hold up this PR, so I opened #335.